openpgp/v2/write.go

```diff
@@ -0,0 +1,1033 @@
+	for _, signature := range s.signatures {
+		if _, err := signature.wrappedHash.Write(data); err != nil {
+			return 0, err
+		}
+	}
+	return len(data), nil
+}
+
+func (s detachSignWriter) Close() error {
+	for _, signature := range s.signatures {
+		err := signature.sig.Sign(signature.h, signature.signer, s.config)
+		if err != nil {
+			return err
+		}
+		err = signature.sig.Serialize(s.signatureWriter)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func detachSignWithWriter(w io.Writer, signers []*Entity, sigType packet.SignatureType, config *packet.Config) (ptWriter io.WriteCloser, err error) {
+	var detachSignContexts []*detachSignContext
+	for _, signer := range signers {
+		signingKey, ok := signer.SigningKeyById(config.Now(), config.SigningKey(), config)
+		if !ok {
+			return nil, errors.InvalidArgumentError("no valid signing keys")
+		}
+		if signingKey.PrivateKey == nil {
+			return nil, errors.InvalidArgumentError("signing key doesn't have a private key")
+		}
+		if signingKey.PrivateKey.Encrypted {
+			return nil, errors.InvalidArgumentError("signing key is encrypted")
+		}
+		candidateHashes := []uint8{
+			hashToHashId(crypto.SHA256),
+			hashToHashId(crypto.SHA384),
+			hashToHashId(crypto.SHA512),
+			hashToHashId(crypto.SHA3_256),
+			hashToHashId(crypto.SHA3_512),
+		}
+		defaultHashes := candidateHashes[0:1]
+		primarySelfSignature, _ := signer.PrimarySelfSignature(config.Now())
+		if primarySelfSignature == nil {
+			return nil, errors.InvalidArgumentError("signed entity has no valid self-signature")
+		}
+		preferredHashes := primarySelfSignature.PreferredHash
+		if len(preferredHashes) == 0 {
+			preferredHashes = defaultHashes
+		}
+		candidateHashes = intersectPreferences(candidateHashes, preferredHashes)
+
+		var hash crypto.Hash
+		if hash, err = selectHash(candidateHashes, config.Hash(), signingKey.PrivateKey); err != nil {
+			return
+		}
+
+		detachSignCtx := detachSignContext{
+			signer: signingKey.PrivateKey,
+		}
+
+		detachSignCtx.sig = createSignaturePacket(signingKey.PublicKey, sigType, config)
+		detachSignCtx.sig.Hash = hash
+
+		detachSignCtx.h, err = detachSignCtx.sig.PrepareSign(config)
+		if err != nil {
+			return
+		}
+		detachSignCtx.wrappedHash, err = wrapHashForSignature(detachSignCtx.h, sigType)
+		if err != nil {
+			return
+		}
+		detachSignContexts = append(detachSignContexts, &detachSignCtx)
+	}
+
+	return &detachSignWriter{
+		signatureWriter: w,
+		signatures:      detachSignContexts,
+		config:          config,
+	}, nil
+}
+
+// FileHints contains metadata about encrypted files. This metadata is, itself,
+// encrypted.
+type FileHints struct {
+	// IsUTF8 can be set to hint that the contents are utf8 encoded data
+	IsUTF8 bool
+	// FileName hints at the name of the file that should be written. It's
+	// truncated to 255 bytes if longer. It may be empty to suggest that the
+	// file should not be written to disk. It may be equal to "_CONSOLE" to
```

Comment:
This comment is outdated

Reply:
3db51732b5b372aebda2697e4cdccbbbf3bb2fec